### DISCUSSION TITLE
Problems using prebuilt python package

### DISCUSSION THREAD

**Discussion by rosshemsley** [InteractiveComputerGraphics/SPlisHSPlasH#151] [Q&A]
Hello,

I installed the pre-built Python package on linux in a virtualenv,

```
$ pip install pySPlisHSPlasH
$ Requirement already satisfied: pySPlisHSPlasH in ./env/lib/python3.8/site-packages (2.11.2)
```

However I get path errors when trying to execute the examples,

```python
In [1]: import pysplishsplash as sph
   ...:

In [2]:     base = sph.Exec.SimulatorBase()
   ...:     base.init(useGui=False)
   ...:
SPlisHSPlasH version: 2.11.2
Output directory: /usr/home/user/.pyenv/versions/3.8.11/bin/output/DoubleDamBreak
Load scene file: /usr/home/user/.pyenv/versions/3.8.11/bin/data/Scenes/DoubleDamBreak.json
Error: Cannot open file!
Initialize static boundary simulation
```

When I try to instead use a custom scene file (I just copied `empty.json` into a local file called `scene.json`) and then build the scene using the other example, I get a different error (the base object has no getScene method).

```python
In [5]: base.init(sceneFile='/usr/local/user/home/user/envs/splishsplash/scene.json')
SPlisHSPlasH version: 2.11.2
Output directory: /usr/local/user/home/user/.pyenv/versions/3.8.11/bin/output/scene
Load scene file: /usr/local/user/home/user/envs/splishsplash/scene.json
Initialize static boundary simulation

In [6]: base.getScene()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [6], in <module>
----> 1 base.getScene()

AttributeError: 'pysplishsplash.Exec.SimulatorBase' object has no attribute 'getScene'
```

Any pointers welcome :)

**Comment by janbender** [comment]
Typically the scene files are installed in a data/Scenes subfolder in your Python directory. Can you check if the directory exists?

**Comment by janbender** [comment]
Btw. how did you start the simulation? By the command "splash"?

**Reply by ssungho98** [reply]
Hi! I also installed the pre-built Python package and faced the same problem. I couldn't use command "splash" after the installation.

**Reply by digitalillusions** [reply]
what OS are you running on? can you check the output of `python -c "import pysplishsplash"`

**Reply by ssungho98** [reply]
#254 I raised this as a separate discussion here!

**Reply by ssungho98** [reply]
When I tried `splash` in pySPlisHSPlasH folder and ran one of the example files, I got the following error:
```
sungho@sungholix:~/SPlisHSPlasH/pySPlisHSPlasH$ splash
Failed to open file: /usr/bin/output/animated_body/log/SPH_log.txt
SPlisHSPlasH version: 2.12.3
Output directory: /usr/bin/output/animated_body
Load scene file: /home/sungho/SPlisHSPlasH/pySPlisHSPlasH/examples/animated_body.py
Error: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'i'
```

**Reply by ssungho98** [reply]
And for `python -c "import pysplishsplash", I got this:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named pysplishsplash
```

**Comment by janbender** [comment]
Do you have multuiple Python installations? It seems that the paths are somehow wrong.

**Reply by ssungho98** [reply]
I think I have both python3.8 and python2.7. I deleted python2.7 and set python3.8 as my default python. Now both `python -c "import pysplishsplash"` and `splash` work.
```
sungho@sungholix:~$ python --version
Python 3.8.10
```
The below message showed up when I tried to open an example file by using `splash`.
```
sungho@sungholix:~/SPlisHSPlasH/pySPlisHSPlasH$ splash
Failed to open file: /usr/bin/output/animated_body/log/SPH_log.txt
SPlisHSPlasH version: 2.12.3
Output directory: /usr/bin/output/animated_body
Load scene file: /home/sungho/SPlisHSPlasH/pySPlisHSPlasH/examples/animated_body.py
Error: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'i'
```

**Comment by janbender** [comment]
I just wonder if your system mixed up both installations. Since you mentioned that you installed the module using pip, but then python did not find the module. So maybe pip was from one Python installation and python from the other one. Just a guess.

**Reply by ssungho98** [reply]
When you said both installations, did you mean both "python" installation or "splishsplash" installation? I first installed Splishs plash through CMake and then installed pysplishsplash through pip. How can I avoid mixing up installations then? What is the proper way of installation?

**Reply by ssungho98** [reply]
Plus, I simply run `~/SPlisHSPlasH/pySPlisHSPlasH/examples/animated_body.py` in VS code to test it. Is this a correct way?

**Comment by janbender** [comment]
Typically, you install pysplishsplash using pip, and start the example using python. pip and python should be from the same Python installation. We do not maintain the code for Python 2.7. So you should use a current Python version.

**Reply by ssungho98** [reply]
I made it work in conda venv. In venv, I could run sample scenes in `~/anaconda3/envs/venv/bin/data/Scenes`. But, I was still not able to run pysplishsplash examples in `~/SPlisHSPlasH/pySPlisHSPlasH/examples`. Could you suggest any method I can import them into conda venv?

**Comment by janbender** [comment]
I really cannot reproduce your problems. Typically you install the module and the examples can load the module. 

Are you able to execute the "SPHSimulator.py" example with the path of a scene file as argument?

**Reply by ssungho98** [reply]
I reinstalled the files in the clean ubuntu environment and then it worked! However, some of the scene files were not operated with the error message
```
Failed to open file: /usr/bin/output/Emitter_2D/log/SPH_log.txt
SPlisHSPlasH version: 2.12.3
Output directory: /usr/bin/output/Emitter_2D
Load scene file: /home/sungho/SPlisHSPlasH/data/Scenes/Emitter_2D.json
Initialize static boundary simulation
Initialize fluid particles
Number of fluid particles: 102
OpenGL version 4.6
Vendor: Intel
Renderer: Mesa Intel(R) Graphics (ADL GT2)
Version: 4.6 (Compatibility Profile) Mesa 21.2.6
Error occurred while loading shader: /usr/bin/resources/shaders/vs_points_vector.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/fs_points.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/vs_points_scalar.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/fs_points.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/vs_points_scalar.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/fs_points_colormap.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/vs_smooth.glsl
Error occurred while loading shader: /usr/bin/resources/shaders/fs_smooth.glsl
```
The same error occurred when I opened them with `splash` command.

**Reply by Samev** [reply]
Ran into the same issue with a setup using `pipenv` and figured I'd make a comment:

I believe the issue is caused by `static std::string getProgramPath()` resolving symlinks.

In my case I've got a `pipenv` environment where I've installed `pysplishsplash`. The installation directory looks good and relevant splishsplash files are located there:
```
ls /home/<redacted>/.local/share/virtualenvs/splishsplash-evaluation-6I07Wa1T/bin -al
total 84
drwxr-xr-x 4 <redacted> <redacted> 4096 Aug 20 17:41 .
drwxr-xr-x 5 <redacted> <redacted> 4096 Aug 20 17:33 ..
-rw-r--r-- 1 <redacted> <redacted> 2325 Aug 20 17:33 activate
-rw-r--r-- 1 <redacted> <redacted> 1592 Aug 20 17:33 activate.csh
-rw-r--r-- 1 <redacted> <redacted> 3155 Aug 20 17:33 activate.fish
-rw-r--r-- 1 <redacted> <redacted> 2840 Aug 20 17:33 activate.nu
-rw-r--r-- 1 <redacted> <redacted> 1696 Aug 20 17:33 activate.ps1
-rw-r--r-- 1 <redacted> <redacted> 1359 Aug 20 17:33 activate_this.py
drwxr-xr-x 4 <redacted> <redacted> 4096 Aug 20 17:41 data
-rwxr-xr-x 1 <redacted> <redacted>  286 Aug 20 17:33 f2py
-rwxr-xr-x 1 <redacted> <redacted>  286 Aug 20 17:33 numpy-config
-rwxr-xr-x 1 <redacted> <redacted>  291 Aug 20 17:33 pip
-rwxr-xr-x 1 <redacted> <redacted>  291 Aug 20 17:33 pip-3.10
-rwxr-xr-x 1 <redacted> <redacted>  291 Aug 20 17:33 pip3
-rwxr-xr-x 1 <redacted> <redacted>  291 Aug 20 17:33 pip3.10
lrwxrwxrwx 1 <redacted> <redacted>   16 Aug 20 17:33 python -> /usr/bin/python3
lrwxrwxrwx 1 <redacted> <redacted>    6 Aug 20 17:33 python3 -> python
lrwxrwxrwx 1 <redacted> <redacted>    6 Aug 20 17:33 python3.10 -> python
drwxr-xr-x 5 <redacted> <redacted> 4096 Aug 20 17:41 resources
-rwxr-xr-x 1 <redacted> <redacted>  301 Aug 20 17:41 splash
-rwxr-xr-x 1 <redacted> <redacted>  278 Aug 20 17:33 wheel
-rwxr-xr-x 1 <redacted> <redacted>  278 Aug 20 17:33 wheel-3.10
-rwxr-xr-x 1 <redacted> <redacted>  278 Aug 20 17:33 wheel3
-rwxr-xr-x 1 <redacted> <redacted>  278 Aug 20 17:33 wheel3.10
```

However since `static std::string getProgramPath()` follows the symbolic link it sets the `SimulatorBase::m_exePath` to `/usr/bin` rather than `/home/<redacted>/.local/share/virtualenvs/splishsplash-evaluation-6I07Wa1T/bin` resulting in not finding e.g. the shaders.
